include/qore/intern/QC_TreeMap.h

```diff
@@ -83,28 +83,43 @@ class TreeMapData : public AbstractPrivateData {
       }
    }
 
-   DLLLOCAL AbstractQoreNode *get(const QoreStringNode *key, ExceptionSink *xsink) const {
+   DLLLOCAL AbstractQoreNode *get(const QoreStringNode *key, const ReferenceNode* unmatched, ExceptionSink *xsink) const {
       TempEncodingHelper keyStr(key, QCS_DEFAULT, xsink);
       if (!keyStr) {
          return 0;
       }
 
       QoreAutoRWReadLocker al(rwl);
-      if (data.empty()) {
-         return 0;
-      }
-
-      std::string path(keyStr->getBuffer());
-
-      Map::const_iterator b = data.begin();
-      Map::const_iterator it = data.upper_bound(path);
-
-      size_t prefixLen = getFirstPathSegmentLength(path);
-      while (it != b && !(--it)->first.compare(0, prefixLen, path, 0, prefixLen)) {
-         if (isPathPrefix(it->first, path)) {
-            return it->second.getReferencedValue();
+      if (!data.empty()) {
+         std::string path(keyStr->getBuffer());
+
+         Map::const_iterator b = data.begin();
+         Map::const_iterator it = data.upper_bound(path);
+
+         size_t prefixLen = getFirstPathSegmentLength(path);
+         while (it != b && !(--it)->first.compare(0, prefixLen, path, 0, prefixLen)) {
+            if (isPathPrefix(it->first, path)) {
+               if (unmatched) {
+                  // it.first.length();
+                  size_t l = it->first.length();
+                  if (isPathEnd(path[l])) {
+                     l++;
+                  }
+                  QoreTypeSafeReferenceHelper ref(unmatched, xsink);
+                  if (!ref)
+                     return 0;
+                  ref.assign(key->substr(l, xsink));
```

Comment:
in general whenever a Qore-language exception is raised, NOTHING should be returned, so this code should more properly look like
```
if (ref.assign(key->substr(l, xsink)))
    return 0;
```